File: src/main/webui/src/model/common/Modifiers.ts
import Modifier from '../common/Modifier';

export default class Modifiers {
    modifiers: Array<Modifier>;
    constructor(modifiers: Modifier[]) {
        this.modifiers = modifiers;
    }
    
    modifier(name: string): Modifier {
        const foundModifier = this.modifiers.find((modifier) => {
            return modifier.name === name;
        });
        if (!foundModifier) {
            throw new Error(`Modifier with name ${name} not found`);
        }
        return foundModifier;
    }
}

File: src/main/java/io/tbbc/cf/modifier/Modifiers.java
package io.tbbc.cf.modifier;

import io.tbbc.cf.common.InternalException;
import io.tbbc.cf.property.CategoryName;
import io.tbbc.cf.property.PropertyName;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

public record Modifiers(List<Modifier> modifiers) {
    public Modifiers(List<Modifier> modifiers) {
        if (modifiers.stream().distinct().count() != modifiers.size()) {
            throw new InternalException("Modifiers must have unique names.");
        }
        this.modifiers = Collections.unmodifiableList(modifiers);
    }

    public List<Modifier> modifiers() {
        return modifiers;
    }

    public Optional<Modifier> modifier(PropertyName name, CategoryName categoryName) {
        return modifiers.stream().filter(modifier -> modifier.name().equals(name)
                || modifier.name().equals(categoryName)).findFirst();
    }
}
